apm-agent-plugins/apm-log-shader-plugin/apm-log-shader-plugin-common/src/main/java/co/elastic/apm/agent/log/shader/AbstractLogShadingHelper.java

```diff
@@ -0,0 +1,146 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2020 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.log.shader;
+
+import co.elastic.apm.agent.configuration.CoreConfiguration;
+import co.elastic.apm.agent.impl.ElasticApmTracer;
+import co.elastic.apm.agent.impl.GlobalTracer;
+import co.elastic.apm.agent.impl.payload.ServiceFactory;
+import co.elastic.apm.agent.impl.transaction.AbstractSpan;
+import co.elastic.apm.agent.logging.LogEcsReformatting;
+import co.elastic.apm.agent.logging.LoggingConfiguration;
+import co.elastic.apm.agent.sdk.state.GlobalState;
+import co.elastic.apm.agent.sdk.weakmap.WeakMapSupplier;
+import com.blogspot.mydailyjava.weaklockfree.WeakConcurrentMap;
+
+import javax.annotation.Nullable;
+
+/**
+ * The abstract Log shading helper- loaded as part of the agent core (agent CL / bootstrap CL / System CL).
+ * Annotated with {@link GlobalState} because it holds the global mapping from original appender to corresponding
+ * shade-appender.
+ *
+ * @param <A> logging-framework-specific Appender type
+ */
+@GlobalState
+public abstract class AbstractLogShadingHelper<A> {
+
+    public static final String ECS_SHADE_APPENDER_NAME = "EcsShadeAppender";
+
+    private static final Object NULL_APPENDER = new Object();
+
+    private final ElasticApmTracer tracer;
+    private final LoggingConfiguration loggingConfiguration;
+    @Nullable
+    private final String configuredServiceName;
+
+    public AbstractLogShadingHelper() {
+        this.tracer = GlobalTracer.requireTracerImpl();
+        loggingConfiguration = tracer.getConfig(LoggingConfiguration.class);
+        configuredServiceName = new ServiceFactory().createService(tracer.getConfig(CoreConfiguration.class), "").getName();
```

Comment:
Why not like this?
```suggestion
        configuredServiceName = tracer.getConfig(CoreConfiguration.class).getServiceName();
```

Reply:
Because it is a bit more future-proof. I want to get here the exact same service name that is reported in the metadata, so if in the future we need to make it all-capital, or append a prefix, or apply any logic instead of only using the configured name as is, it will still remain in sync.

Reply:
Got it. I think to make it even more future-proof and to solve the issue about which service name to use, we should expose a `ElasticApmTracer#getServiceName(ClassLoader)` method. In contrast to what the existing private method does, it should get the service name from the `MetaData` if there's no override for the provided class loader. We can probably get the class loader of the application by calling `Thread.currentThread().getContextClassLoader()`.

One complication is that the metadata is inside a future, but that should be resolvable as the service name doesn't depend on the cloud metadata.

Another issue is that the service name might not be known that early in the application startup. For example, the detection of the `spring.application.name` happens after the first event has been logged. But at least for spring boot, there seems to be a fix. We can instrument either `org.springframework.boot.SpringApplication#prepareEnvironment` or `org.springframework.boot.context.config.ConfigFileApplicationListener#onApplicationEvent`, or similar which allows us to read `environment.getProperty("spring.application.name")` before the first event is logged.

Reply:
We decided to start with leaving it as it is now and see how we enhance it later on. In any case, when the user is setting the `service_name` explicitly, then it will work as expected.